File: lintcode/MinimumWindowSubstring.cpp
class Solution {
public:    
    /**
     * @param source: A string
     * @param target: A string
     * @return: A string denote the minimum window
     *          Return "" if there is no such a string
     */
    string minWindow(string &source, string &target) {
        unordered_map<char, int> toFind, found;
        for (auto c : target) toFind[c] ++;
        
        int total = 0, minLen = -1, minStart = -1;
        for (int end = 0, begin = 0; end < source.size(); ++ end) {
            char e = source[end];
            if (toFind[e] == 0) continue;
            found[e] ++;
            if (found[e] <= toFind[e]) ++ total;
            
            if (total == target.length()) {
                char b = source[begin];
                while (toFind[b] == 0 || found[b] > toFind[b]) { // note this!
                    if (found[b] > toFind[b]) -- found[b];
                    b = source[++ begin];
                }
                
                if (minLen == -1 || minLen > end - begin + 1) {
                    minLen = end - begin + 1;
                    minStart = begin;
                }
            }
        }
        
        return minLen == -1 ? "" : source.substr(minStart, minLen);
    }
};


/**
Minimum Window Substring

Given a string source and a string target, find the minimum window in source 
which will contain all the characters in target.

Notice

If there is no such window in source that covers all characters in target, 
return the emtpy string "".

If there are multiple such windows, you are guaranteed that there will 
always be only one unique minimum window in source.
Clarification

Should the characters in minimum window has the same order in target?

    Not necessary.

Example

For source = "ADOBECODEBANC", target = "ABC", the minimum window is "BANC"
Challenge

Can you do it in time complexity O(n) ?

 */

File: leetcode_java/MinimumWindowSubstring.java
class Solution {
    public String minWindow(String s, String t) {
        int m = s.length(), n = t.length();
        if (m < n) return "";

        HashMap<Character, Integer> toFind = new HashMap<>();
        HashMap<Character, Integer> found = new HashMap<>();
        
        for (int i = 0; i < n; ++ i) {
            char c = t.charAt(i);
            toFind.put(c, toFind.get(c) == null ? 1 : (toFind.get(c) + 1));
        }
        
        int total = 0, minLen = 0, pos = 0;
        for (int i = 0, start = 0; i < m; ++ i) {
            char c = s.charAt(i);
            found.put(c, found.get(c) == null ? 1 : (found.get(c) + 1));
            if (toFind.get(c) == null) continue;
            if (found.get(c) <= toFind.get(c)) total += 1;
            
            while (total == n) {
                char d = s.charAt(start);
                if (toFind.get(d) == null || found.get(d) > toFind.get(d)) {
                    ++ start;
                    found.put(d, found.get(d) - 1);
                } else {
                    int len = i - start + 1;
                    if (minLen == 0 || minLen > len) {
                        minLen = len;
                        pos = start;
                    }
                    break;
                }
            }
        }
        return s.substring(pos, pos + minLen);
    }
}

/**
Minimum Window Substring

Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that 
every character in t (including duplicates) is included in the window. If there is no such substring, return 
the empty string "".
 */
